### ISSUE TITLE
Future integration

### ISSUE THREAD

**Issue by nappa85** [Amanieu/parking_lot#155]
Hello,
I really like your job, I use this crate really a lot, and I found the lack of a Future integration quite sad, because I often write future-based software and using locks inside them isn't a good choice.

So I spent a bit of my free time and I came up with this really simple (and quite incomplete) implementation: [https://github.com/nappa85/future-rwlock/blob/master/src/lib.rs](https://github.com/nappa85/future-rwlock/blob/master/src/lib.rs)

I don't know if you had other plans about this topic, maybe you'll find my work horrible, but I wanted to share my two cents.

Best regards

**Comment by Amanieu** [comment]
Unfortunately Futures integration requires a completely different and incompatible API for locks, at which point it may as well just be a completely separate crate.

My position is that this is outside the scope of parking_lot. See #24 and #86.

**Comment by nappa85** [comment]
Yes, I already read those issues, my implementation (I made some more edits since I opened this issue, maybe you missed it) is some kind different, it works on anything that implements AsRef<RwLock<T>> (e.g. Arc, Rc and Box) with obvious limitations about Send and Sync.
This way, core RwLock remains untouched.
The same approach could be applied also to Mutex.

**Comment by Amanieu** [comment]
I read through your code, and overall it seems pretty good. One significant issue is that `task::current().notify()` should only be called once the mutex is unlocked by a thread, rather than every time the future is polled. This results in 100% CPU usage as tokio will keep polling the mutex (with `try_lock`) in a loop.

Unfortunately this is not possible to solve without deeper integration in the internals of the mutex.

**Comment by nappa85** [comment]
Yes, I understand your point, I was aware of the problem (it was described in one of the previous issues).
I used the crate in some real implementations, the CPU usage doesn't rise because futures solve so quickly that a good percent never task::notify once and the remaining task::notify only once.
I think it's better than having futures with blocking code inside, but it can be done better, no dubts on it.

**Comment by faern** [comment]
If you are happy with a solution that is basically just a busy spin loop, then you don't need `parking_lot` at all. You don't get any of the benefits of `parking_lot` if I understand your usage correctly.

**Comment by nappa85** [comment]
> If you are happy with a solution that is basically just a busy spin loop, then you don't need `parking_lot` at all. You don't get any of the benefits of `parking_lot` if I understand your usage correctly.

Sorry, so a "busy spin loop" that 90% of the times turns only once is worst than blocking code inside futures?

**Comment by faern** [comment]
A spin loop based lock that never have to actually spin, that can proceed through the lock on the first or second try every time is indeed very efficient. But that's in theory. If you write a generic library you don't know what programmers will do while they hold the lock, right? So if they do anything that hold the lock for some time all other futures will completely clog up the CPU. I would personally not implement such a library unless the documentation made it very clear locks must only be held for an extremely short period of time. And probably not then either since your current design breaks the intended usage of futures:

```rust
task::current().notify();
return Ok(Async::NotReady)
```
^ You first notify the waker that your task has more data/can proceed, but then you at the same time return that you ran out of work (`NotReady`). This is a contradiction.

And no, you should probably not block inside futures neither. That will block all futures running on that thread inside that runtime/reactor. It goes against the fundamental intended usage of futures.

In conclusion, you can't do what you want to do in a good way using the existing lock design. As @Amanieu points out, you need locks with a completely different design made for an async future use case.

You need locks that upon polling a locked lock registers an interest in that lock with the task currently holding it. And then the task holding the lock must `notify()` those interested in it when it's unlocked.

**Comment by nappa85** [comment]
Yes, thank you, it was the "can be done better" part, but, as far I can see, it's not quick and easy, or someone else would have already done it.
As you pointed out, I don't know what users will do inside the lock, but it's in the basis of futures themselves not doing long and heavy operations, because it would block the runtime/reactor thread, exactly like using blocking software.

About the portion of code you cited, how can I continue the loop (aka returning NotReady) and notify the task he has to poll again?

**Comment by faern** [comment]
> About the portion of code you cited, how can I continue the loop (aka returning NotReady) and notify the task he has to poll again?

The point is you should not. By returning `NotReady` and at the same time telling the runtime to poll you again you basically say "I return now, please call me again asap", and you are not supposed to do that. This is by definition creating a CPU eating busy loop.

**Comment by faern** [comment]
You likely want to look at something like this: https://crates.io/crates/qutex

The point is that future aware locks are incompatible to system thread parking/`parking_lot`. You likely never want thread parking synchronization primitives in any async context at all. The fundamental difference between `parking_lot` waiting for a lock and waiting for something (`foo`) in an async context is:
* `parking_lot` tells the OS to park the thread, pausing execution until `foo` happens.
* an async task tells the runtime "Please don't schedule/poll me again until `foo` happens".

Basically both "pauses" a thread. The difference is that one is telling the OS scheduler to pause the OS thread and the other is telling the user space runtime to pause a green thread/task. In the latter case the OS thread executing the user space runtime can continue executing other green threads/tasks until `foo` happens.

The problem you have here is that you have no way of telling the runtime to poll you when `foo` happens (the lock is unlocked), because `parking_lot` does not expose any way for you to get this information.

**Comment by nappa85** [comment]
Using the new async/await syntax I've been able to do something like you were talking about.
The approach has changed, now the system revolves around a wrapper on the RawLock (RawMutex and RawRwLock) and retains an internal list of wakers.
When a Future asks for a lock and can't obtain it, the Future's Waker is added to the list.
As soon as the lock is unlocked, the wakers are waked up.

At the moment, ALL wakers are waked up.
I'm studing if there are corner cases where waking up only the first waker has some negative behaviour, to light up the process.

**Comment by faern** [comment]
I still don't think you want to use thread locking in an async program/algorithm at all. Why do you want to do this? If you block a thread you will prohibit other async tasks on that thread from running.

**Comment by nappa85** [comment]
Where do you see thread locking in my code?
It's the purpose of it having Mutex and RwLock non-threadlocking

**Comment by 0xpr03** [comment]
There are multiple things here, maybe my comment can help to clarify this:
- If you stay on a single threaded system reactor you normally don't ever have to use locks. Because you (hopefully) "drop" your mutex the moment you enter the next async awaiting code (and thus returning execution to tokio). Thus you can use a Rc if required at all.
- If you do reach a point of a mutex not being available and you're using a multi-threaded reactor (and not simply having an actor-based approach where you have message passing for single-instance Objects), you do not want to block at all. "Parking" the thread would mean that you park your whole reactor thread and thus any other futures currently in the pipeline of your reactor (tokio). What you really want to do is to tell tokio that you currently can't further execute this task (like if you'd wait for IO operation) and let it execute another future. Then the current lock holder (thread) would have to issue a notify call to tokio afterwards when it release the lock, resuming the execution if it is currently halted. (Which happens when all current Futures in the reactor are in "waiting" mode, thus the reactor parks itself).
This has one big difference to parking lot: In parking lot you park the whole thread of the caller that wants to acquire a lock. But you never do that in async code. In async code only the future executor (tokio) is allowed to "park" the thread as only it knows if there is nothing to do anymore.

Thus parking_lot is a completely different concept than the one you need for a futures aware synchronization system. `Non-Threadlocking` is the opposite of what parking_lot is designed for.
